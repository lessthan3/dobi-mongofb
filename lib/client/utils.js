"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.startsWith = exports.prepareFind = exports.jsonify = exports.log = exports.isEquals = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var isEquals = function isEquals(a, b) {
  if (a && !b) {
    return false;
  }

  if (b && !a) {
    return false;
  }

  if ((0, _typeof2.default)(a) !== (0, _typeof2.default)(b)) {
    return false;
  }

  if (a === null && b === null) {
    return true;
  }

  switch ((0, _typeof2.default)(a)) {
    case 'function':
      if (a.toString() !== b.toString()) {
        return false;
      }

      break;

    case 'object':
      if (Object.keys(a).length !== Object.keys(b).length) {
        return false;
      }

      for (var k in a) {
        if (!isEquals(a[k], b[k])) {
          return false;
        }
      }

      break;

    default:
      if (a !== b) {
        return false;
      }

  }

  return true;
}; // logging utility


exports.isEquals = isEquals;

var log = function log(msg) {
  return console.log("[monogfb] ".concat(msg));
}; // stringify json params


exports.log = log;

var jsonify = function jsonify(q) {
  var o = {};

  var _arr = Object.keys(q);

  for (var _i = 0; _i < _arr.length; _i++) {
    var k = _arr[_i];
    var v = q[k];

    if (v) {
      o[k] = JSON.stringify(v);
    }
  }

  return o;
}; // prepare query parameters for a find


exports.jsonify = jsonify;

var prepareFind = function prepareFind() {
  var fields;
  var next;
  var options;
  var special;

  for (var _len = arguments.length, _args = new Array(_len), _key = 0; _key < _len; _key++) {
    _args[_key] = arguments[_key];
  }

  var args = _args.filter(function (arg) {
    return arg != null;
  }); // callback


  var hasCallback = typeof args[args.length - 1] === 'function';

  if (hasCallback) {
    next = args[args.length - 1];
  } // defaults


  var criteria = {}; // query objects

  if ((0, _typeof2.default)(args[0]) === 'object') {
    var _args2 = (0, _slicedToArray2.default)(args, 1);

    criteria = _args2[0];
  }

  if ((0, _typeof2.default)(args[1]) === 'object') {
    var _args3 = (0, _slicedToArray2.default)(args, 2);

    fields = _args3[1];
  }

  if ((0, _typeof2.default)(args[2]) === 'object') {
    var _args4 = (0, _slicedToArray2.default)(args, 3);

    options = _args4[2];
  }

  if ((0, _typeof2.default)(args[3]) === 'object') {
    var _args5 = (0, _slicedToArray2.default)(args, 4);

    special = _args5[3];
  } // args[1] can be either fields or options or special
  // args[2] can be either options or special
  // case: special was in args[2]


  if (options && !special && (options.token || options._)) {
    var _ref = [options, null];
    special = _ref[0];
    options = _ref[1];
  } // case: options was in args[1]


  if (fields && !options && (fields.limit || fields.skip || fields.sort)) {
    var _ref2 = [fields, null];
    options = _ref2[0];
    fields = _ref2[1];
  } // case: special was in args[1]


  if (fields && !special && (fields.token || fields._)) {
    var _ref3 = [fields, null];
    special = _ref3[0];
    fields = _ref3[1];
  } // format query objects and prepare to send


  var query = {
    criteria: criteria,
    fields: fields,
    options: options
  };
  var params = jsonify(query);

  if (special != null ? special.token : undefined) {
    params.token = special.token;
  }

  if (special != null ? special._ : undefined) {
    params._ = special._;
  }

  return [query, params, next];
};

exports.prepareFind = prepareFind;

var startsWith = function startsWith(str, target) {
  return str.slice(0, target.length) === target;
};

exports.startsWith = startsWith;