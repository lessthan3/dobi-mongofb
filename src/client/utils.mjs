import isEqualWith from 'lodash/isEqualWith';

const customIsEqual = (a, b) => {
  if (typeof a === 'function' && typeof b === 'function') {
    return a.toString() === b.toString();
  }
  return undefined;
};

export const isEqual = (a, b) => (
  isEqualWith(a, b, customIsEqual)
);

// stringify json params
export const jsonify = (q) => {
  const o = {};
  for (const [k, v] of Object.entries(q)) {
    if (v) {
      o[k] = JSON.stringify(v);
    }
  }
  return o;
};

// prepare query parameters for a find
export const prepareFind = (..._args) => {
  let fields;
  let next;
  let options;
  let special;

  const args = _args.filter(arg => arg != null);

  // callback
  const hasCallback = typeof args[args.length - 1] === 'function';
  if (hasCallback) {
    next = args[args.length - 1];
  }

  // defaults
  let criteria = {};

  // query objects
  if (typeof args[0] === 'object') {
    [criteria] = args;
  }
  if (typeof args[1] === 'object') {
    [, fields] = args;
  }
  if (typeof args[2] === 'object') {
    [, , options] = args;
  }
  if (typeof args[3] === 'object') {
    [, , , special] = args;
  }

  // args[1] can be either fields or options or special
  // args[2] can be either options or special

  // case: special was in args[2]
  if (options && !special && (options.token || options._)) {
    [special, options] = [options, null];
  }

  // case: options was in args[1]
  if (fields && !options && (
    fields.limit || fields.skip || fields.sort
  )) {
    [options, fields] = [fields, null];
  }

  // case: special was in args[1]
  if (fields && !special && (
    fields.token || fields._
  )) {
    [special, fields] = [fields, null];
  }

  // format query objects and prepare to send
  const query = { criteria, fields, options };
  const params = jsonify(query);

  if (special && special.token) {
    params.token = special.token;
  }
  if (special && special._) {
    params._ = special._;
  }

  return [query, params, next];
};

export const startsWith = (str, target) => str.slice(0, target.length) === target;
