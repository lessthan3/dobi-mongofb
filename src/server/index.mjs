/* eslint-disable no-underscore-dangle */

// dependencies
import assert from 'assert';
import Router from 'koa-router';
import middleware from './middleware';

import {
  find,
  findOne,
  insert,
  read,
  remove,
  update,
} from './routes';

const {
  findByIdMiddleware,
  findMiddleware,
  findOneMiddleware,
  insertMiddleware,
  removeMiddleware,
  updateMiddleware,
} = middleware;

/**
 *
 * @param {Object} config
 * @param {Object} config.cache
 * @param {boolean} config.cache.enabled=true
 * @param {number} config.cache.max=100
 * @param {string} config.cache.redisUri=localhost
 * @param {Object[]} config.firebaseShards
 * @param {string} config.firebaseShards.*.apiKey
 * @param {Object} config.firebaseShards.*.credential
 * @param {string} config.firebaseShards.*.databaseURL
 * @param {Object} config.middleware
 * @param {Object} config.mongodb
 * @param {Object} config.options
 * @param {string[]} config.options.collections=[] array of collections
 * @param {number} config.options.limitDefault=20 default limit of query response
 * @param {number} config.options.limitMax=1000 max returned results
 * @param {boolean} config.options.setCreated=true
 * @param {boolean} config.options.setLastModified=true
 * @param {function} config.validateSync
 * @param {string} config.root='/api' endpoint root
 */
export default (config) => {
  // configuration
  const { root } = config;
  assert(root, 'root required');
  assert(root[0] === '/', 'root must start with /');

  // build routes
  const router = new Router();
  router.get(`${root}/:collection/find`, findMiddleware, find);
  router.get(`${root}/:collection/findOne`, findOneMiddleware, findOne);
  router.delete(`${root}/:collection/:id`, removeMiddleware, remove);
  router.get(`${root}/:collection/:id/*`, findByIdMiddleware, read);
  router.get(`${root}/:collection/:id`, findByIdMiddleware, read);
  router.patch(`${root}/:collection/:id`, updateMiddleware, update);
  router.post(`${root}/:collection`, insertMiddleware, insert);
  return router;
};
