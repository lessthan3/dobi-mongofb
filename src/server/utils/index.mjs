export { default as cleanObject } from './cleanObject';
export const invalidFirebaseKeyChars = /[$#[\]\/\s{}]/;
