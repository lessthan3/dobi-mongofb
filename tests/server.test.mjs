import request from 'supertest';
import bodyParser from 'koa-bodyparser';
import Koa from 'koa';
import hasIn from 'lodash/hasIn';
import set from 'lodash/set';
import mongodb from 'mongodb';
import mongoFb from '../src/server';
import baseConfig from './config';
import { createFirebaseMock, createMongoMock } from './utils';
import { cleanObject } from '../src/server/utils';
import {
  alphaPage,
  alphaPageId,
  alphaSite,
  alphaSiteId,
  alphaUserId,
  betaSiteId,
  betaUserId,
  newPageId,
} from './mockDocuments';

const { root } = baseConfig;
const { ObjectId } = mongodb;

// server
let app;

const {
  adminMock,
  refMock,
  setMock,
} = createFirebaseMock(jest);

const {
  deleteOneMock,
  findMock,
  findOneMock,
  insertOneMock,
  mongodbMock,
  updateMock,
} = createMongoMock(jest);

beforeEach(() => {
  const server = new Koa();
  server.use(bodyParser());
  server.use(mongoFb({
    ...baseConfig,
    testModules: { admin: adminMock, mongodb: mongodbMock },
  }));
  app = server.callback();
});

afterEach(() => {
  jest.clearAllMocks();
});

describe.only('GET :collection/findOne', () => {
  it('returns expected values', async () => {
    const response = await request(app).get(`${root}/sites/findOne`)
      .query({
        criteria: JSON.stringify(({
          _id: alphaSiteId,
          'users.test-alphaAdmin@test,com': 'admin',
        })),
      });

    // expect(response.text).toBe('hahha');
    expect(response.status).toBe(200);
    expect(response.body).toEqual({ ...alphaSite, _id: alphaSiteId });
    expect(findOneMock).toHaveBeenCalledWith('sites', {
      'users.test-alphaAdmin@test,com': 'admin',
    });
  });
});

describe('utils/cleanObject', () => {
  it('parses primary values correctly', () => {
    const obj = {
      array: [
        { foo: 'bar' },
        ['foo', null, undefined, [], { emptyArray: [], emptyObject: {} }],
        0,
        '',
        'foo',
        null,
        undefined,
        {},
        [],
        [[], {}],
        { emptyArray: [], emptyObject: [] },
      ],
      arrayEmpty: [null, undefined, [[], {}], { emptyArray: [], emptyObject: [] }],
      objectEmpty: {
        nullKey: null,
        undefinedKey: undefined,
      },
      booleanTrue: true,
      booleanFalse: false,
      object: {
        array: ['foo'],
        arrayEmpty: [null, undefined, [[], {}], { emptyArray: [], emptyObject: [] }],
        booleanTrue: true,
        booleanFalse: false,
        numberFloat: 123.45,
        numberInt: 123,
        numberZero: 0,
        object: { foo: 'bars' },
        objectEmpty: {
          nullKey: null,
          undefinedKey: undefined,
        },
        string: 'foo',
        stringEmpty: '',
      },
      numberFloat: 123.45,
      numberInt: 123,
      numberZero: 0,
      string: 'foo',
      stringEmpty: '',
    };
    const result = cleanObject(obj);
    expect(result.array).toMatchSnapshot();
    expect(hasIn(result, 'arrayEmpty')).not.toBeTruthy();
    expect(hasIn(result, 'objectEmpty')).not.toBeTruthy();
    expect(result.booleanTrue).toBe(true);
    expect(result.booleanFalse).toBe(false);
    expect(result.object).toMatchSnapshot();
    expect(result.numberFloat).toMatchSnapshot();
    expect(result.numberInt).toEqual(123);
    expect(result.numberInt).toMatchSnapshot();
    expect(result.numberZero).toMatchSnapshot();
    expect(result.string).toMatchSnapshot();
    expect(result.stringEmpty).toMatchSnapshot();
  });
});

describe('DELETE :collection/:id', () => {
  it('fails on bad collection', async () => {
    const response = await request(app)
      .delete(`${root}/fakeCollection/${alphaSiteId}`);
    expect(response.status).toBe(404);
  });

  it('fails if no auth', async () => {
    const response = await request(app)
      .delete(`${root}/pages/${alphaPageId}`);
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });

  it('fails if no auth', async () => {
    const response = await request(app)
      .delete(`${root}/pages/${alphaPageId}`);
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });

  it('fails if user not admin (client middleware test)', async () => {
    const response = await request(app)
      .delete(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaUser')
      .set('x-bearer-token-shard', 'testAlpha');
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });

  it('returns if user is admin (client middleware test)', async () => {
    const response = await request(app)
      .delete(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha');
    expect(response.status).toBe(200);
    expect(refMock).toHaveBeenCalledWith(`pages/${alphaPageId}`);
    expect(setMock).toHaveBeenCalledWith(null);
    expect(setMock).toHaveBeenCalledTimes(2);
    expect(deleteOneMock).toHaveBeenCalledWith('pages', { _id: new ObjectId(alphaPageId) });
  });
});

describe('GET :collection/:id', () => {
  it('fails on bad collection', async () => {
    const response = await request(app)
      .get(`${root}/fakeCollection/${alphaSiteId}`);
    expect(response.status).toBe(404);
  });
  it('fails on bad id', async () => {
    const response = await request(app)
      .get(`${root}/sites/123`);
    expect(response.status).toBe(400);
    expect(response.text).toMatchSnapshot();
  });
  it('fails on protected value if no auth', async () => {
    const response = await request(app)
      .get(`${root}/users/${alphaUserId}`);
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });
  it('fails on protected value if wrong user', async () => {
    const response = await request(app)
      .get(`${root}/users/${alphaUserId}`)
      .set('Authorization', 'Bearer betaUser')
      .set('x-bearer-token-shard', 'testAlpha');
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });
  it('fails on protected value if wrong admin (client middleware test)', async () => {
    const response = await request(app)
      .get(`${root}/users/${betaUserId}`)
      .set('Authorization', 'Bearer alphaUser')
      .set('x-bearer-token-shard', 'testAlpha');
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });
  it('returns on protected value when authd', async () => {
    const response = await request(app)
      .get(`${root}/users/${alphaUserId}`)
      .set('Authorization', 'Bearer alphaUser')
      .set('x-bearer-token-shard', 'testAlpha');
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
    expect(findOneMock).toHaveBeenCalledWith('users', {
      _id: new ObjectId(alphaUserId),
    });
  });
  it('returns on protected value when user is admin (client middleware test)', async () => {
    const response = await request(app)
      .get(`${root}/users/${alphaUserId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha');
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
    expect(findOneMock).toHaveBeenCalledWith('users', {
      _id: new ObjectId(alphaUserId),
    });
  });
  it('returns expected value', async () => {
    const response = await request(app)
      .get(`${root}/sites/${alphaSiteId}`);
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
    expect(findOneMock).toHaveBeenCalledWith('sites', {
      _id: new ObjectId(alphaSiteId),
    });
  });
});

describe('PATCH :collection/:id', () => {
  let mockWrite;
  let mockKey;
  let mockValue;
  let mockValueClean;
  beforeEach(() => {
    mockKey = 'data.test';
    mockValue = { item: 'myUpdate', author: { name: 'steve', extra: [{}, [], null] } };
    mockValueClean = { item: 'myUpdate', author: { name: 'steve' } };
    mockWrite = { key: mockKey, value: { ...mockValue } };
  });
  it('fails if user not authd', async () => {
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .send(mockWrite);
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if bad collection', async () => {
    const response = await request(app)
      .patch(`${root}/fakeCollection/${alphaPageId}`)
      .send(mockWrite);
    expect(response.status).toBe(404);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if user not admin (client middleware test)', async () => {
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer betaUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send(mockWrite);
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if user doesnt send a value', async () => {
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({ key: mockKey });
    expect(response.status).toBe(400);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if key isnt a string', async () => {
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({ key: 5, value: { ...mockValue } });
    expect(response.status).toBe(400);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if key is empty (meaning root) and value isnt an object', async () => {
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({ key: '', value: [] });
    expect(response.status).toBe(400);
    expect(response.text).toMatchSnapshot();
  });
  for (const char of ['$', '#', '[', ']', '/', ' ', '  ', '{', '}']) {
    it(`fails if key is contains '${char}'`, async () => {
      const response = await request(app)
        .patch(`${root}/pages/${alphaPageId}`)
        .set('Authorization', 'Bearer alphaAdminUser')
        .set('x-bearer-token-shard', 'testAlpha')
        .send({ key: `data.${char}`, value: { ...mockValue } });
      expect(response.status).toBe(400);
      expect(response.text).toMatchSnapshot();
    });
  }
  it('fails if key part contains empty string', async () => {
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({ key: 'data..thing', value: { ...mockValue } });
    expect(response.status).toBe(400);
    expect(response.text).toMatchSnapshot();
  });

  it('fails if a root write but the _id was updated', async () => {
    const newId = '5bdd46a0b966ff619846ae1b';
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({ key: '', value: { ...mockValue, _id: newId } });
    expect(response.status).toBe(400);
    expect(response.text).toMatchSnapshot();
  });

  it('succeeds for non-root update if admin is authed (client middleware test)', async () => {
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({ key: mockKey, value: { ...mockValue } });
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
    expect(refMock).toHaveBeenCalledWith(`pages/${alphaPageId}/data/test`);
    expect(setMock).toHaveBeenCalledWith({ ...mockValueClean });
    expect(setMock).toHaveBeenCalledTimes(2);
    expect(updateMock).toHaveBeenCalledWith('pages', {
      _id: new ObjectId(alphaPageId),
    }, {
      ...set({ ...alphaPage }, mockKey, mockValueClean),
      _id: new ObjectId(alphaPageId),
    });
  });

  it('succeeds for root update if admin is authed (client middleware test)', async () => {
    const rootUpdate = set({ ...alphaPage }, mockKey, mockValue);
    const rootUpdateClean = set({ ...alphaPage }, mockKey, mockValueClean);
    const firebaseSetValue = { ...rootUpdateClean, _id: rootUpdateClean._id.toString() };
    const response = await request(app)
      .patch(`${root}/pages/${alphaPageId}`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({ key: '', value: { ...rootUpdate } });
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
    expect(refMock).toHaveBeenCalledWith(`pages/${alphaPageId}`);
    expect(setMock).toHaveBeenCalledWith(firebaseSetValue);
    expect(updateMock).toHaveBeenCalledWith('pages', {
      _id: new ObjectId(alphaPageId),
    }, {
      ...rootUpdateClean,
      _id: new ObjectId(alphaPageId),
    });
  });
});

describe('POST :collection/:id', () => {
  let mockValue;
  beforeEach(() => {
    mockValue = {
      created: 200,
      name: 'gammaPage',
      site_id: betaSiteId,
    };
  });

  it('fails if user not authd', async () => {
    const response = await request(app)
      .post(`${root}/users`)
      .send({ value: { ...mockValue } });
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if bad collection', async () => {
    const response = await request(app)
      .post(`${root}/fakeCollection`)
      .send({ value: { ...mockValue } });
    expect(response.status).toBe(404);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if user doesnt send a value', async () => {
    const response = await request(app)
      .post(`${root}/pages`)
      .set('Authorization', 'Bearer alphaUser')
      .set('x-bearer-token-shard', 'testAlpha');
    expect(response.status).toBe(400);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if object has an _id in it', async () => {
    const response = await request(app)
      .post(`${root}/pages`)
      .set('Authorization', 'Bearer alphaUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({
        value: { ...mockValue, _id: newPageId },
      });
    expect(response.status).toBe(400);
    expect(response.text).toMatchSnapshot();
  });
  it('fails if user isnt an admin (client middleware test)', async () => {
    const response = await request(app)
      .post(`${root}/pages`)
      .set('Authorization', 'Bearer alphaUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({ value: { ...mockValue } });
    expect(response.status).toBe(401);
    expect(response.text).toMatchSnapshot();
  });

  it('returns if user is admin (client middleware test)', async () => {
    const response = await request(app)
      .post(`${root}/pages`)
      .set('Authorization', 'Bearer alphaAdminUser')
      .set('x-bearer-token-shard', 'testAlpha')
      .send({
        value: {
          created: 200,
          name: 'gammaPage',
          site_id: alphaSiteId,
        },
      });
    expect(response.status).toBe(200);
    expect(response.text).toMatchSnapshot();
    expect(insertOneMock).toHaveBeenCalledWith('pages', {
      created: 200,
      name: 'gammaPage',
      site_id: alphaSiteId,
    });
    expect(refMock).toHaveBeenCalledWith(`pages/${newPageId}`);
    expect(setMock).toHaveBeenCalledTimes(2);
    expect(setMock).toHaveBeenCalledWith({
      _id: newPageId,
      created: 200,
      name: 'gammaPage',
      site_id: alphaSiteId,
    });
  });
});
