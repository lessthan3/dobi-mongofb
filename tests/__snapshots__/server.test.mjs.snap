// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE :collection/:id fails if no auth 1`] = `"Unauthorized"`;

exports[`DELETE :collection/:id fails if no auth 2`] = `"Unauthorized"`;

exports[`DELETE :collection/:id fails if user not admin (client middleware test) 1`] = `"unauthorized to remove"`;

exports[`GET :collection/:id fails on bad id 1`] = `"invalid id"`;

exports[`GET :collection/:id fails on protected value if no auth 1`] = `"Unauthorized"`;

exports[`GET :collection/:id fails on protected value if wrong admin (client middleware test) 1`] = `"Unauthorized"`;

exports[`GET :collection/:id fails on protected value if wrong user 1`] = `"Unauthorized"`;

exports[`GET :collection/:id returns expected value 1`] = `
Object {
  "_id": "5bdce894a1649c5974fb30c3",
  "created": 100,
  "name": "alphaSite",
  "users": Object {
    "test-alphaAdmin@test,com": "admin",
  },
}
`;

exports[`GET :collection/:id returns on protected value when authd 1`] = `
Object {
  "_id": "5bdce8afa1649c5974fb30c4",
  "created": 100,
  "site_id": "5bdce894a1649c5974fb30c3",
  "uid": "test-alpha@test.com",
}
`;

exports[`GET :collection/:id returns on protected value when user is admin (client middleware test) 1`] = `
Object {
  "_id": "5bdce8afa1649c5974fb30c4",
  "created": 100,
  "site_id": "5bdce894a1649c5974fb30c3",
  "uid": "test-alpha@test.com",
}
`;

exports[`PATCH :collection/:id fails if a root write but the _id was updated 1`] = `"modifying _id not permitted"`;

exports[`PATCH :collection/:id fails if bad collection 1`] = `"Not Found"`;

exports[`PATCH :collection/:id fails if key is contains '  ' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is contains ' ' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is contains '#' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is contains '$' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is contains '/' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is contains '[' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is contains ']' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is contains '{' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is contains '}' 1`] = `"invalid characters in key"`;

exports[`PATCH :collection/:id fails if key is empty (meaning root) and value isnt an object 1`] = `"update value must be object if modifying root"`;

exports[`PATCH :collection/:id fails if key isnt a string 1`] = `"invalid key"`;

exports[`PATCH :collection/:id fails if key part contains empty string 1`] = `"cannot contain empty part in key (e.g. foo..bar)"`;

exports[`PATCH :collection/:id fails if user doesnt send a value 1`] = `"missing value"`;

exports[`PATCH :collection/:id fails if user not admin (client middleware test) 1`] = `"unauthorized to update"`;

exports[`PATCH :collection/:id fails if user not authd 1`] = `"Unauthorized"`;

exports[`PATCH :collection/:id succeeds for non-root update if admin is authed (client middleware test) 1`] = `
Object {
  "author": Object {
    "name": "steve",
  },
  "item": "myUpdate",
}
`;

exports[`PATCH :collection/:id succeeds for root update if admin is authed (client middleware test) 1`] = `
Object {
  "_id": "5bdcea2ba5b86e5a65b09bb1",
  "created": 100,
  "data": Object {
    "test": Object {
      "author": Object {
        "name": "steve",
      },
      "item": "myUpdate",
    },
  },
  "name": "alphaPage",
  "site_id": "5bdce894a1649c5974fb30c3",
}
`;

exports[`POST :collection/:id fails if bad collection 1`] = `"Not Found"`;

exports[`POST :collection/:id fails if object has an _id in it 1`] = `"value must not contain _id"`;

exports[`POST :collection/:id fails if user doesnt send a value 1`] = `"invalid value to insert"`;

exports[`POST :collection/:id fails if user isnt an admin (client middleware test) 1`] = `"unauthorized to insert"`;

exports[`POST :collection/:id fails if user not authd 1`] = `"Unauthorized"`;

exports[`POST :collection/:id returns if user is admin (client middleware test) 1`] = `"{\\"created\\":200,\\"name\\":\\"gammaPage\\",\\"site_id\\":\\"5bdce894a1649c5974fb30c3\\",\\"_id\\":\\"5bdd2f98b966ff619846ae1a\\"}"`;

exports[`utils/cleanObject parses primary values correctly 1`] = `
Array [
  Object {
    "foo": "bar",
  },
  Array [
    "foo",
  ],
  0,
  "",
  "foo",
]
`;

exports[`utils/cleanObject parses primary values correctly 2`] = `
Object {
  "array": Array [
    "foo",
  ],
  "booleanFalse": false,
  "booleanTrue": true,
  "numberFloat": 123.45,
  "numberInt": 123,
  "numberZero": 0,
  "object": Object {
    "foo": "bars",
  },
  "string": "foo",
  "stringEmpty": "",
}
`;

exports[`utils/cleanObject parses primary values correctly 3`] = `123.45`;

exports[`utils/cleanObject parses primary values correctly 4`] = `123`;

exports[`utils/cleanObject parses primary values correctly 5`] = `0`;

exports[`utils/cleanObject parses primary values correctly 6`] = `"foo"`;

exports[`utils/cleanObject parses primary values correctly 7`] = `""`;
